### ISSUE TITLE
Thread field updates should be handled by a single PATCH route

### ISSUE THREAD

**Issue by jnaviask** [hicommonwealth/commonwealth#4664]
## Description

Currently, our routes used to modify threads are split across many different routes, for historical reasons. As we move toward Common as platform, we should unify all the various routes that update thread state into a more clear and concise API, which will simplify both how our 1st party application and 3rd party application developers interface with the platform.

### API's / Server

Our current list of thread-modifying routes is as follows (per @mzparacha in #4588):

**APIs for fetching threads**

- `GET /threads` -- does not need changes
- `GET /linking/getLinks` - used to get links on threads - may want to expose a `GET /thread/:id/links` in addition to a `/linking` domain for now, although in the future we may have links be a standalone feature. The reason is that we may want to fetch links on a proposal page rather than on a threads page, so the API should be agnostic in that regard.

**APIs for updating thread fields**

Note the permissions listed after the route name.

- `POST /linking/addThreadLinks` (author, admin, mod) - for adding a link thread/discussion to an existing thread 
- `DELETE /linking/deleteLinks` (author, admin, mod) - for removing a link thread/discussion to an existing thread
    - May want to rename to `POST /thread/:id/links` and `DELETE /thread/:id/links`, but also potentially out of scope for this ticket.

The following updates we may want to combine into a single PATCH call, permissioned as noted:

- `POST /updateThreadPrivacy` (author, admin, mod) --  used to lock/unlock a thread.
- `POST /updateThreadStage` (author, admin, mod) -- used to update thread stage.
- `POST /updateTopic` (admin) -- used to update thread topic.
- `POST /updateThreadPinned` (admin, mod) -- used to update thread pinned status.
- `POST /threads/${threadId}/spam` (admin, mod -- currently author can mark, incorrectly) -- used to update thread spam status.
- `POST /addEditors` (author -- currently anyone can mark, incorrectly) -- used to add collaborators to thread
- `POST /deleteEditors` (author) -- used to removed collaborators from thread
- `PATCH /threads/${threadId}` (???) -- used to update a thread title/description and basic info

### Frontend
- Update all frontend API calls with the updated response 

Please provide additional details as needed to properly scope ticket, then we can proceed after #4588 is shipped.

**Comment by mzparacha** [comment]
What about array fields like `thread.reactions`. To update this via `PUT/PATCH`, we would need to pass the full reactions array I assume?

If that's the case I think we should make an assumption that the frontend only knows the "new/updated" data and not the current one. And we keep dedicated endpoints for anything array-based in the thread object.

Thoughts?

cc: @rbennettcw @jnaviask

**Comment by rbennettcw** [comment]
I don't see a need for reactions to be changed from the current routes:
- `POST /threads/:id/reactions`
- `POST /comments/:id/reactions`
- `DELETE /reactions/:id`

---

Regarding what's mentioned in the ticket, the only thing I think is fuzzy is how we'll handle editors (also an array). For that, I see two options:
- A: Client sends the entire array of editors to overwrite the existing value
- B: Client sends a payload with operations like so:
```ts
{
  "add": [...],
  "remove": [...]
}
```

Option A is simpler, but also more destructive– a frontend bug could wipeout all of the existing editors.

**Comment by mzparacha** [comment]
Included frontend scope as well and assigned myself as well

cc: @jnaviask @rbennettcw

**Comment by mzparacha** [comment]
> I don't see a need for reactions to be changed from the current routes:
POST /threads/:id/reactions
POST /comments/:id/reactions
DELETE /reactions/:id

Apologies, I didn't mean that. I was speaking in favor of doing it tho 😄

> If that's the case I think we should make an assumption that the frontend only knows the "new/updated" data and not the current one. And we keep dedicated endpoints for anything array-based in the thread object.

+1, Totally in favor of Option B, kudos
> B: Client sends a payload with operations like so:


cc: @rbennettcw

**Comment by rbennettcw** [comment]
Bumping to 5 since each operation has different permissions. I'm working on a clean solution to implement that.
